### PR TITLE
GPG-387 (GPG-524 and 526) Update "Manage Organisations" Pages

### PR DESCRIPTION
**Pull request by dulciejackson**
The first chunk of functionality for the ManageOrganisations journey!

This:
- Creates a new ManageOrganisationsController
- Creates new ManageOrganisationsViewModel and ManageOrganisationViewModel
- Implements functionality for ManageOrganisations and ManageOrganisation GET endpoints
- Creates replacement views with new design system styling
- Adds feature flag for Launchpad.cshtml redirect on "submit gender pay gap data" link
- Adds OrganisationReportsSection.cshtml to replace DisplayReportSection with new design system styling
- Adds UsersRegisteredToReportForOrganisation.cshtml to separate out the ManageOrganisation.cshtml file a bit (in keeping with the above reports table)

### CODE REVIEW

File: GenderPayGap.WebUI/Controllers/Guidance/EightWaysGuidanceController.cs
Comment (dulciejackson) [inline GenderPayGap.WebUI/Controllers/Guidance/EightWaysGuidanceController.cs:77]: IntelliJ automatically did this - I think it updated the characters used for the single quotes (even though it doesn't look like it)!

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml
Comment (dulciejackson) [inline GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml:66]: I wasn't sure if this chunk of C# was better than the old version (in `Views/Organisation/ManageOrganisations.cshtml`) but it separated it out a bit from lots of processing alongside lots of html?

File: GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml
Comment (dulciejackson) [inline GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml:20]: I feel like this file is horrible - it feels like a real code smell to me that there's _so much_ C# alongside the HTML here. I tried to refactor some of the complexity out, but I think a lot of it is pretty much essential so this uses the right one of the many complicated variable urls/bits of text in each place!

File: GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs
Comment (dulciejackson) [reply GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs:35]: As far as I can tell, this isn't used elsewhere at the moment - although it seems slightly odd to me that we'd only show the privacy policy here and not elsewhere? 

@jamesgriff do you think this would be useful?

File: GenderPayGap.WebUI/Models/Organisation/OrganisationReportsViewModel.cs
Comment (Binney) [inline GenderPayGap.WebUI/Models/Organisation/OrganisationReportsViewModel.cs:36]: There's a few comments here I find overkill - no particular downside to having them but I definitely don't think they're needed and there's a case for deleting them just to keep the lines in the file down

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisation.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisation.cshtml:33]: Yup, I copied and pasted too fast!

File: GenderPayGap.WebUI/Models/Organisation/OrganisationReportsViewModel.cs
Comment (Binney) [inline GenderPayGap.WebUI/Models/Organisation/OrganisationReportsViewModel.cs:152]: Does this need to be here? can you expose a getter on `snapshotDateForYear` directly?

File: GenderPayGap.WebUI/Views/Viewing/Launchpad/Index.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/Viewing/Launchpad/Index.cshtml:42]: I think we'd normally look through uses of the feature flag and refactor to point to just the "new" version, and then remove the feature flag entirely :)

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml
Comment (Binney) [inline GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml:50]: I'm deeply confused about how Github is showing me this diff but these appear to still be present and not be needed any more

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml
Comment (Binney) [inline GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml:98]: `n`?

File: GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml
Comment (Binney) [inline GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml:16]: Since this is just a `foreach`, could we just do it as a list of `OrganisationScopeForYear`? (would also make the generation of Model.OrganisationScopesByYear a little easier)

File: GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml
Comment (Binney) [inline GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml:21]: This should also go in the viewmodel

File: GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml
Comment (Binney) [inline GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml:64]: Github is complaining about these, which I think is inadvertently because it expects each HTML attribute to be on only one line - thoughts? i.e. a line break before the `href`

File: GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml
Comment (Binney) [inline GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml:24]: Could this go in the viewmodel too? (Instead of having a local var at all I mean - making `GetByReportingDeadlineText` take no arguments)

File: GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs
Comment (Binney) [reply GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs:41]: I like it! I agree with Dulcie, it does feel odd we only show the privacy policy here - but I think that's a problem with the business logic not with the code, so pulling it out feels right to me

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisation.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisation.cshtml:26]: Ok, I just went with what the design system suggested for this sort of case - is there a reason the admin site doesn't use that styling?

File: GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs
Comment (dulciejackson) [reply GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs]: Clever!
Is it semantically "good practice" for an exception to redirect to an actual page rather than an error page of some kind? Not questioning it, just curious!

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisation.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisation.cshtml:54]: This has been bundled into a whole grouping of other design work (e.g. Manage Organisation/Account tabs) that Claire's going to get to at some point and then feed into more specific dev tickets. For now it's been agreed with Jacob to build this exactly as it is right now to stop this being blocked by the need for more design!

File: GenderPayGap.WebUI/Views/Viewing/Launchpad/Index.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/Viewing/Launchpad/Index.cshtml:44]: I've added a feature flag check in each of the old and new manage organisations endpoints to redirect to the other one if the feature flag is enabled/disabled respectively - so a user can't manually input the new journey URL when it's not meant to be live and vice versa. [GPG-517](https://technologyprogramme.atlassian.net/browse/GPG-527) is in the pipeline to swap all other links over to pointing at the right one (although in practice I don't think it'll make a difference if we're checking the feature flag in both directions)

File: GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs
Comment (dulciejackson) [reply GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs:32]: Do we want to give a 404, or just automatically redirect to the "right" version of the page? I imagine it'd be confusing to e.g. follow a link to Manage Organisations, it have swapped to the new version on live, and you get a 404 rather than the controller leading to the right place!

File: GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml:16]: How I didn't think of this, I have no idea 😅

File: GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs
Comment (Binney) [reply GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs]: Well I sort of like "exception as fall-through behaviour" - notwithstanding it's sort of the only way we'd do this particular thing - an exception shouldn't just be "something has gone wrong" but "something has happened that's an exception to the usual rules of what is going on" - no reason that behaviour should be pathological. This is quite similar to "redirect to a signin page if you're not logged in", certainly it's a pattern we use across BBC Monitoring (where the permissions system is multi-layered)

File: GenderPayGap.WebUI/Models/Organisation/OrganisationReportsViewModel.cs
Comment (Binney) [inline GenderPayGap.WebUI/Models/Organisation/OrganisationReportsViewModel.cs:34]: Fun fact: you can do this in a super functional way using Linq (e.g. `reportingYears.Select(organisation.GetScopeForYear)`, filter on not-nulls, then Select to create the OrganisationScopeForYear) but this is probably clearer if we're not doing things the functional way elsewhere

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisation.cshtml
Comment (Binney) [inline GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisation.cshtml:15]: crumbs

File: GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs
Comment (Binney) [inline GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs:65]: (I've lost track of what's already on this branch/review/PR but) I think there's a helper for this (the "404 if decryption fails")

**Review by Binney — APPROVED** [review]
Niceeeeeee

File: GenderPayGap.WebUI/Models/Organisation/OrganisationReportsViewModel.cs
Comment (dulciejackson) [reply GenderPayGap.WebUI/Models/Organisation/OrganisationReportsViewModel.cs:34]: Ah yeah, I love these functional Linq extras!
I'm really keen on functional stuff, but aware that that's not classic C# style and might be a bit unfamiliar to people with a C# background picking up maintenance in future. Also aware that other places use `foreach`, so trying to keep it consistent! :)

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisation.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisation.cshtml:15]: 🍞

File: GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs
Comment (jamesgriff) [reply GenderPayGap.WebUI/Controllers/ManageOrganisations/ManageOrganisationsController.cs:32]: For these pages, your solution works nicely.
For other pages, I think a 404 would be appropriate (e.g. if it's a new feature where the public shouldn't know that the feature exists yet).
It can sometimes get a bit messy trying to work out what the best equivalent page is (e.g. if the structure of the journey has changed) so sometimes a 404 is cleanest - but it depends on the particular case.
I think this redirect-one-way-or-the-other works nicely here.

File: GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml
Comment (jamesgriff) [reply GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml]: I like the order:
```cshtml
<a href="..."
   class="govuk-link">
    Link text
</a>
```
It helps me remember that there should always be a `class="govuk-link"` on each `<a>` tag (except in this case where it's a button)

Comment (jamesgriff) [reply GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml]: Actually, styling links as buttons is frowned upon ☹️ for accessibility reasons (happy to have a call to discuss more details)
If these are going to be buttons, then it might be better if each one is inside a form:
```cshtml
<form method="get"
      action="@(Url.Action(...))">
    <button .../>
</form>
```
I know that's a bit of a bore, but it does genuinely help people using speech-to-command software

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml]: Yeah, this was me mis-typing on impulse! The original (and designs) use full-width, so I've reverted back to that and it should look better now!

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml:102]: I'm not sure what to do about the conditional bit of text here; the "If you created an account by mistake or no longer responsible for reporting gender pay gap data you can close your account..." - I'm not quite sure how to determine when to show it?

**Review by jamesgriff — COMMENTED** [review]
I've addressed some of my comments in https://github.com/cabinetoffice/gender-pay-gap/pull/268

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml
Comment (Binney) [inline GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml:95]: Do we have a way of tracking TODOs anywhere? All else being equal this should at least tag the ticket number i.e. GPG-387

File: GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml]: Could the `aria-role` come in handy here? That way we could keep the links, style them as buttons but a screen-reader would present them as buttons?

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml
Comment (dulciejackson) [reply GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml:95]: Ah this is a TODO until I figure out how to do it/if it's necessary in this PR! Just there to remind myself to come back to before merging, but I doubt it warrants a whole new ticket :)

File: GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml
Comment (jamesgriff) [reply GenderPayGap.WebUI/Views/ManageOrganisations/OrganisationReportsTable.cshtml]: Ooh yes, I think that would work - great thought 😄

File: GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml
Comment (jamesgriff) [reply GenderPayGap.WebUI/Views/ManageOrganisations/ManageOrganisations.cshtml:102]: I think this will be
```cshtml
@if (Model.UserOrganisations.Count == 0)
{
    ..."If you created your account"...
}
```

File: GenderPayGap.WebUI/Models/Organisation/OrganisationReportsViewModel.cs
Comment (dulciejackson) [reply GenderPayGap.WebUI/Models/Organisation/OrganisationReportsViewModel.cs]: I've added a ticket to refactor this into a helper method and update all references!

**Comment by jamesgriff** [comment]
@dulciejackson 
I've added a few commits - can you review?
Other than that, I'm happy for it to be merged 😄

**Review by jamesgriff — APPROVED** [review]

**Pull request merged**
